File: deleteWithoutHead.java
// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes (GFG)
// Any problem you faced while coding this : None

/*
 * Approach:
 * Since the head is not given, we replace the given node's data with the data
 * of the next node (including it's link), and "remove" the node to the next of
 * the given node.
 */

/*
class Node
{
	int data ;
	Node next;
	Node(int d)
	{
		data = d;
		next = null;
	}
}
*/

//Function to delete a node without any reference to head pointer.
class Solution
{
    void deleteNode(Node del)
    {
        // Your code here
        if(del.next != null) {
            del.data = del.next.data;
            del.next = del.next.next;
        }
    }
}

File: deletewithouthead.go
// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
if node.Next == nil set node as nil
else copy next element value to current and set current.Next = next element.next
*/
package main

import "fmt"

func deleteNode(node *ListNode) {
	if node.Next == nil {
		node = nil
	} else {
		node.Val = node.Next.Val
		node.Next = node.Next.Next
	}
}

func MainDelete() {
	head:= &ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 3, Next: &ListNode{Val: 4, Next: &ListNode{Val: 5}}}}}
	deleteNode(head.Next.Next)
	for head != nil{
		fmt.Println(head.Val)
		head = head.Next
	}// expected 1->2->4->5
}